ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/encoding/TransactionDecoder.java

```diff
@@ -14,38 +14,47 @@
  */
 package org.hyperledger.besu.ethereum.core.encoding;
 
-import static com.google.common.base.Preconditions.checkNotNull;
-
 import org.hyperledger.besu.datatypes.TransactionType;
 import org.hyperledger.besu.ethereum.core.Transaction;
 import org.hyperledger.besu.ethereum.rlp.RLP;
 import org.hyperledger.besu.ethereum.rlp.RLPInput;
 
 import java.util.Optional;
 
-import com.google.common.collect.ImmutableMap;
 import org.apache.tuweni.bytes.Bytes;
 
 public class TransactionDecoder {
 
   @FunctionalInterface
-  interface Decoder {
+  public interface Decoder {
     Transaction decode(RLPInput input);
   }
 
-  private static final ImmutableMap<TransactionType, Decoder> TYPED_TRANSACTION_DECODERS =
-      ImmutableMap.of(
-          TransactionType.ACCESS_LIST,
-          AccessListTransactionDecoder::decode,
-          TransactionType.EIP1559,
-          EIP1559TransactionDecoder::decode,
-          TransactionType.BLOB,
-          BlobTransactionDecoder::decode,
-          TransactionType.DELEGATE_CODE,
-          CodeDelegationTransactionDecoder::decode);
+  @FunctionalInterface
+  public interface DecoderProvider {
+    /**
+     * Gets the decoder for a given transaction type and encoding context. If the context is
+     * POOLED_TRANSACTION, it uses the network decoder for the type. Otherwise, it uses the typed
+     * decoder.
+     *
+     * @param transactionType the transaction type
+     * @param encodingContext the encoding context
+     * @return the decoder
+     */
+    Decoder getDecoder(TransactionType transactionType, EncodingContext encodingContext);
+  }
+
+  /** The decoder provider. Defaults to the MainnetTransactionEncoderDecoderProvider. */
+  private static DecoderProvider decoderProvider = new MainnetTransactionEncoderDecoderProvider();
 
-  private static final ImmutableMap<TransactionType, Decoder> POOLED_TRANSACTION_DECODERS =
-      ImmutableMap.of(TransactionType.BLOB, BlobPooledTransactionDecoder::decode);
+  /**
+   * Sets the decoder provider to be used for decoding transactions.
+   *
+   * @param decoderProvider the decoder provider to be used
+   */
+  public static void setDecoderProvider(final DecoderProvider decoderProvider) {
```

Comment:
Seems ok for now, ideally we would provide the decoder provider via dagger and/or plugin services, but we can wait until we do the plugin implementation to sort that out.